src/arch/xen/xen_memory.c

```diff
@@ -0,0 +1,317 @@
+               "Truncating memory from %luMB to ",
+               ((unsigned long)pfn_to_virt(*max_pfn) - VIRT_START)>>20);
+        */
+        *max_pfn = virt_to_pfn(HYPERVISOR_VIRT_START - PAGE_SIZE());
+        /*
+        printk("%luMB\n",
+               ((unsigned long)pfn_to_virt(*max_pfn) - VIRT_START)>>20);
+        */
+    }
+
+
+    start_address = (unsigned long)pfn_to_virt(pfn_to_map);
+    end_address = (unsigned long)pfn_to_virt(*max_pfn);
+
+    /* We worked out the virtual memory range to map, now mapping loop */
+    //printk("Mapping memory range 0x%lx - 0x%lx\n", start_address, end_address);
+
+    while ( start_address < end_address )
+    {
+        tab = pt_base;
+        pt_mfn = pfn_to_mfn(virt_to_pfn(pt_base));
+
+
+        offset = l3_table_offset(start_address);
+        /* Need new L2 pt frame */
+        if ( !(tab[offset] & _PAGE_PRESENT) )
+            new_pt_frame(&pt_pfn, pt_mfn, offset, L2_FRAME);
+
+        page = tab[offset];
+        pt_mfn = pte_to_mfn(page);
+        tab = to_virt(mfn_to_pfn(pt_mfn) << PAGE_SHIFT);
+        offset = l2_table_offset(start_address);        
+
+        /* Need new L1 pt frame */
+        if ( !(tab[offset] & _PAGE_PRESENT) )
+            new_pt_frame(&pt_pfn, pt_mfn, offset, L1_FRAME);
+
+        page = tab[offset];
+        pt_mfn = pte_to_mfn(page);
+        tab = to_virt(mfn_to_pfn(pt_mfn) << PAGE_SHIFT);
+        offset = l1_table_offset(start_address);
+
+        if ( !(tab[offset] & _PAGE_PRESENT) )
+        {
+            mmu_updates[count].ptr =
+                ((pgentry_t)pt_mfn << PAGE_SHIFT) + sizeof(pgentry_t) * offset;
+            mmu_updates[count].val =
+                (pgentry_t)pfn_to_mfn(pfn_to_map) << PAGE_SHIFT | L1_PROT;
+            count++;
+        }
+        pfn_to_map++;
+        if ( count == L1_PAGETABLE_ENTRIES ||
+             (count && pfn_to_map == *max_pfn) )
+        {
+            rc = HYPERVISOR_mmu_update(mmu_updates, count, NULL, DOMID_SELF);
+            if ( rc < 0 )
+            {
+                /*
+                printk("ERROR: build_pagetable(): PTE could not be updated\n");
+                printk("       mmu_update failed with rc=%d\n", rc);
+                */
+                do_exit();
+            }
+            count = 0;
+        }
+        start_address += PAGE_SIZE();
+    }
+
+    *start_pfn = pt_pfn;
+}
+#if 0
+/*
+ * Clear some of the bootstrap memory
+ */
+static void clear_bootstrap(void)
+{
+    pte_t nullpte = { };
+    int rc;
+
+    /* Use first page as the CoW zero page */
+    memset((void*)VIRT_START, 0, PAGE_SIZE());
+    mfn_zero = virt_to_mfn(VIRT_START);
+
+    if ( (rc = HYPERVISOR_update_va_mapping(0, nullpte, UVMF_INVLPG)) )
+        printk("Unable to unmap NULL page. rc=%d\n", rc);
+}
+#endif
+
+#define MAX_MEM_SIZE            0x3f000000ULL
+void arch_init_mm(unsigned long* start_pfn_p, unsigned long* max_pfn_p)
+{
+    unsigned long start_pfn, max_pfn;
```

Comment:
codestyle:
* function opening `{` should be on the same line as the function last argument
* identation should be a tab (not 4 spaces as here and not 8 spaces)